File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/validation/YearRangeValidator.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.validation;

import java.time.Year;

import lombok.NoArgsConstructor;

/**
 * A Validator that asserts that a year value is within
 * an acceptable range (between 1910 and current year).
 * 
 * This validator checks that years are:
 * - Not before 1910
 * - Not after the current year
 * 
 * If the year is outside this range, the validation fails.
 */
@NoArgsConstructor
public final class YearRangeValidator implements SerializableAttributeValidator {

    private static final long serialVersionUID = 1L;

    private static final int MIN_YEAR = 1910;

    /**
     * Validates that the year value is within acceptable range.
     * 
     * @param value the year string to validate
     * @return true if the year is within acceptable range (1910 to current year),
     *         false otherwise
     */
    @Override
    public boolean eval(String value) {
        if (value == null || value.trim().isEmpty()) {
            return false;
        }

        try {
            int year = Integer.parseInt(value.trim());
            int currentYear = Year.now().getValue();
            return year >= MIN_YEAR && year <= currentYear;
        } catch (NumberFormatException e) {
            // If the value cannot be parsed as an integer, it's invalid
            return false;
        }
    }
}

File: lib/python/opentoken/src/main/opentoken/attributes/validation/year_range_validator.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from datetime import date
from opentoken.attributes.validation.serializable_attribute_validator import SerializableAttributeValidator


class YearRangeValidator(SerializableAttributeValidator):
    """
    A validator that asserts that a year value is within an acceptable range.

    This validator checks that years are:
    - Not before 1910
    - Not after the current year

    If the year is outside this range, the validation fails.
    """

    MIN_YEAR = 1910

    def __init__(self):
        """Initialize the validator with default year range (1910-current year)."""
        pass

    @property
    def max_year(self) -> int:
        """Get the maximum allowed year (current year)."""
        return date.today().year

    def eval(self, value: str) -> bool:
        """
        Validate that the year value is within acceptable range.

        Args:
            value: The year string to validate

        Returns:
            True if the year is within acceptable range (1910-current year),
            False otherwise
        """
        if not value or not value.strip():
            return False

        try:
            year = int(value.strip())
            return self.MIN_YEAR <= year <= self.max_year
        except ValueError:
            return False
